Cpp_hands_on_programming/lib/pipe/Ipc_pipe.cc

```diff
@@ -35,55 +50,103 @@ void Ipc_pipe::send() {
 			throw(std::runtime_error("Ipc_pipe::send: readFile"));
 		}
 		// write to pipe
-		if ((bytes_write = write(fdp, buffer.data(), bytes_read)) == -1) {
-	    	throw(std::runtime_error("Ipc_pipe::send: write to pipe"));
-		}
+		t.startTimer();   
+		do {
+			errno = 0;
+			signal(SIGPIPE, SIG_IGN); // ignore SIGPIPE in case read end is closed 
+			bytes_write = write(fdp, buffer.data(), bytes_read);
+			if (bytes_write == -1) {
+				if (errno == EAGAIN) {
+					t.checkTimer("Pipe is full for too long. Errno: " + std::string(strerror(errno)));
+				} else {
+					throw(std::runtime_error("Ipc_pipe::send: write. Errno: " + std::string(strerror(errno))));
+				}
+			}
+		} while (errno == EAGAIN);
 		// compare read and written bytes 
 		if (bytes_read == bytes_write) {
 			total += bytes_read;
 		} else {
 			throw(std::runtime_error("Ipc_pipe::send: read write difference"));
 		}
-	}
-
-	close(fdp);
+	}  
+ 	close(fdp);
 	std::cout << "File delivered by pipe successfully, exiting the program.." << std::endl;
 }
 
 void Ipc_pipe::receive() {
     std::cout << "Starting pipeReceive.." << std::endl;
 	int fdp, bytes;
-	off_t fileSize(0), total(0);
+	off_t fileSize(0), total(0), sendFileSize(0);
+	bool headerSent(false);
 	std::vector<char> buffer(PIPE_BUF);
 	// open pipe 
-	while ((fdp = open(name.c_str(), O_RDONLY)) == -1) {
-		std::this_thread::sleep_for (std::chrono::seconds(1));
-	}
+	t.startTimer();
+	do {
+		errno = 0;
+		fdp = open(name.c_str(), O_RDONLY | O_NONBLOCK); 
+		t.checkTimer("Cannot connect to pipeSend. Errno: " + std::string(strerror(errno)), "Connecting pipeSend..");
+	} while(errno == ENOENT);
+
+	std::cout << "Connected!" << std::endl;
 	// Start read and write loop
-	while ((bytes = read(fdp, buffer.data(), PIPE_BUF)) != 0) {	
-		if (bytes == -1) {
-			throw(std::runtime_error("Ipc_pipe::receive: read"));
-		}
+	do {
+		t.startTimer();
+		do {
+			errno = 0;
+			bytes = read(fdp, buffer.data(), PIPE_BUF);
+			if (bytes == -1) {
+				if (errno == EAGAIN) {
+					t.checkTimer("Pipe is empty for too long. Errno: " + std::string(strerror(errno)));
+				} else {
+					throw(std::runtime_error("Ipc_pipe::receive: read. Errno: " + std::string(strerror(errno))));
+				}
+			}
+			else if (!headerSent) {
+				if (bytes == 0) {
+					t.checkTimer("Takes too long for write end to open");
+				} else {
+					auto sendFileName = std::string(buffer.begin(), buffer.begin() + bytes);
+					if ((sendFileSize = getFileSize(sendFileName)) == -1) {
+						throw(std::runtime_error("Ipc_pipe::receive: getFilesize-sendFile. Errno: " + std::string(strerror(errno))));
+					}
+					headerSent = true;
+					total -= bytes;
```

Comment:
I don't understand this line. Why are you doin total -= bytes ?

Reply:
The idea is to make sure the header is not written into the new file, and total is reset to 0 after the first loop

Reply:
But in this loop, I didn't see total changed except that line. And maybe if you want to do it, it could be better to write total=0, isn't it?

Reply:
At the end of the outer loop, there is a line of total += bytes, that would offset the change for the first loop. Total is 0 at the beginning, the idea is to make condition total >= 0 for the writing process work and  skip the header

Reply:
Oh Ok I see, sorry.

(in my implementation I get out of the loop the part to receive the header, that's why I was confused :D )